fili-core/src/test/groovy/com/yahoo/bard/webservice/web/apirequest/FilterOptimizingRequestMapperSpec.groovy

```diff
@@ -0,0 +1,155 @@
+
+        when:
+        mapper.apply(request, Mock(ContainerRequestContext))
+
+        then:
+        1 * dim.optimizeFilters([inFilter]) >> expectedFilters
+        1 * request.withFilters(new ApiFilters([(dim): expectedFilters]))
+    }
+
+    def "Optimizations that remove all filters remove the dimension from the filter set as well"() {
+        setup:
+        Dimension dim = Mock(FilterOptimizingDimension)
+        Set<ApiFilter> filters = [Mock(ApiFilter)]
+        DataApiRequest request = Mock(DataApiRequest) { getApiFilters() >> new ApiFilters([(dim): filters] as Map<Dimension, Set<ApiFilter>>)}
+
+        when:
+        mapper.apply(request, Mock(ContainerRequestContext))
+
+        then:
+        1 * dim.optimizeFilters(filters) >> []
+        1 * request.withFilters({ApiFilters it -> it.isEmpty()})
+    }
+
+
+    def "If an optimization adds a new dimension to the filter set the new dimension is properly split out into the map"() {
+        setup:
+        Dimension dim1 = Mock(FilterOptimizingDimension)
+        Dimension dim2 = Mock(FilterOptimizingDimension)
+        ApiFilter dim1OutputFilter = Mock(ApiFilter) {getDimension() >> dim1}
+        ApiFilter dim2OutputFilter = Mock(ApiFilter) {getDimension() >> dim2}
+
+        DataApiRequest request = Mock(DataApiRequest) {getApiFilters() >> new ApiFilters([
+                (dim1): [Mock(ApiFilter)] as Set<ApiFilter>
+        ] as Map<Dimension, Set<ApiFilter>>)}
+
+        when:
+        mapper.apply(request, Mock(ContainerRequestContext))
+
+        then:
+        1 * dim1.optimizeFilters(_) >> [dim1OutputFilter, dim2OutputFilter]
+        0 * dim2.optimizeFilters(_)
+        1 * request.withFilters({
+            ApiFilters it ->
+                it.size() == 2 &&
+                        it.containsKey(dim1) &&
+                        it.containsKey(dim2) &&
+                        it.get(dim1) == [dim1OutputFilter] as Set<ApiFilter> &&
+                        it.get(dim2) == [dim2OutputFilter] as Set<ApiFilter>
+        })
+    }
+
+    def "If an optimization totally transforms a filter into a filter on a totally different dimension, the original dimension is not in the result"() {
+        setup:
+        Dimension dim1 = Mock(FilterOptimizingDimension)
+        Dimension dim2 = Mock(FilterOptimizingDimension)
+        ApiFilter dim2OutputFilter = Mock(ApiFilter) {getDimension() >> dim2}
+
+        DataApiRequest request = Mock(DataApiRequest) {
+            getApiFilters() >> new ApiFilters([
+                    (dim1): [Mock(ApiFilter)]
+            ] as Map<Dimension, Set<ApiFilter>>)
+        }
+
+        when:
+        mapper.apply(request, Mock(ContainerRequestContext))
+
+        then:
+        1 * dim1.optimizeFilters(_) >> [dim2OutputFilter]
+        0 * dim2.optimizeFilters(_)
+        1 * request.withFilters(new ApiFilters([
+                  (dim2) : [dim2OutputFilter] as Set
+            ] as Map<Dimension, Set<ApiFilter>>)
+        )
+    }
```

Comment:
Add a test where if dimension A optimizes into dimension B that the resulting query has the merge of filters for dimension B.  Ensure this happens regardless of the order of the dimensions.